src/integ_test/java/games/strategy/engine/lobby/server/login/LobbyLoginValidatorTest.java

```diff
@@ -117,4 +147,112 @@ public void testLogin() {
+        Util.createUniqueTimeStamp(), mac, address));
+  }
+
+  @Test
+  public void testLegacyLoginCombined() {
+    final LobbyLoginValidator validator = new LobbyLoginValidator();
+    final SocketAddress address = new InetSocketAddress(5000);
+    final String name = Util.createUniqueTimeStamp();
+    final String mac = MacFinder.getHashedMacAddress();
+    final String email = "none@none.none";
+    final String password = "foo";
+    final String hashedPassword = MD5Crypt.crypt(password);
+    new DbUserController().createUser(
+        new DBUser(
+            new DBUser.UserName(name),
+            new DBUser.UserEmail(email)),
+        new HashedPassword(hashedPassword));
+    final Map<String, String> properties = new HashMap<>();
+    final Map<String, String> challengeProperties = validator.getChallengeProperties(name, address);
+    properties.put(LobbyLoginValidator.HASHED_PASSWORD_KEY, hashedPassword);
+    properties.put(LobbyLoginValidator.LOBBY_VERSION, LobbyServer.LOBBY_VERSION.toString());
+    properties.put(RsaAuthenticator.ENCRYPTED_PASSWORD_KEY, RsaAuthenticator
+        .encryptPassword(challengeProperties.get(RsaAuthenticator.RSA_PUBLIC_KEY), password));
+    assertEquals(challengeProperties.get(LobbyLoginValidator.SALT_KEY),
+        MD5Crypt.getSalt(MD5Crypt.MAGIC, hashedPassword));
+    assertNull(new LobbyLoginValidator().verifyConnection(challengeProperties, properties, name, mac, address));
+    assertTrue(BCrypt.checkpw(Util.sha512(password), new DbUserController().getPassword(name).value));
+    properties.remove(RsaAuthenticator.ENCRYPTED_PASSWORD_KEY);
+    assertNull(new LobbyLoginValidator().verifyConnection(challengeProperties, properties, name, mac, address));
+    assertEquals(hashedPassword, new DbUserController().getLegacyPassword(name).value);
+  }
+
+  @Test
+  public void testTimeout() {
+    final LobbyLoginValidator validator = new LobbyLoginValidator();
+    final SocketAddress address = new InetSocketAddress(5000);
+    final String name = Util.createUniqueTimeStamp();
+    final String mac = MacFinder.getHashedMacAddress();
+    final String email = "none@none.none";
+    final String password = "foo";
+    new DbUserController().createUser(
+        new DBUser(
+            new DBUser.UserName(name),
+            new DBUser.UserEmail(email)),
+        Util.sha512(password));
+    final Map<String, String> properties = new HashMap<>();
+    RsaAuthenticator.setTimeout(1, TimeUnit.SECONDS);
+    final Map<String, String> challengeProperties = validator.getChallengeProperties(name, address);
+    properties.put(RsaAuthenticator.ENCRYPTED_PASSWORD_KEY,
+        RsaAuthenticator.encryptPassword(challengeProperties.get(RsaAuthenticator.RSA_PUBLIC_KEY), password));
+    properties.put(LobbyLoginValidator.LOBBY_VERSION, LobbyServer.LOBBY_VERSION.toString());
+    //Wait 1 second for timeout
+    ThreadUtil.sleep(1000);
```

Comment:
Please no sleeps in the test code. Solve by extracting the module that requests the current time, you can then control that and not need real time to go by.